File: preorderTraversal/preorderTraversal.cpp
#include <iostream>
#include <vector>
using namespace std;

// Định nghĩa cấu trúc của một nút trong cây nhị phân
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    // Hàm duyệt cây theo thứ tự trước
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;
        preorder(root, result);
        return result;
    }
    
    // Hàm đệ quy để duyệt cây theo thứ tự trước
    void preorder(TreeNode* node, vector<int>& result) {
        if (node == NULL) return;
        result.push_back(node->val); // Thêm giá trị của nút vào danh sách kết quả
        preorder(node->left, result); // Duyệt qua cây con bên trái
        preorder(node->right, result); // Duyệt qua cây con bên phải
    }
};

int main() {
    // Tạo cây nhị phân tương ứng với ví dụ 1
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(3);

    Solution sol;
    vector<int> result = sol.preorderTraversal(root);

    cout << "Preorder traversal: ";
    for (int num : result) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: preorderTraversal/preorderTraversal.py
from typing import List

# Định nghĩa cấu trúc của một nút trong cây nhị phân
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # Hàm duyệt cây theo thứ tự trước
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        self.preorder(root, result)
        return result
    
    # Hàm đệ quy để duyệt cây theo thứ tự trước
    def preorder(self, node: TreeNode, result: List[int]):
        if not node:
            return
        result.append(node.val) # Thêm giá trị của nút vào danh sách kết quả
        self.preorder(node.left, result) # Duyệt qua cây con bên trái
        self.preorder(node.right, result) # Duyệt qua cây con bên phải

# Tạo cây nhị phân tương ứng với ví dụ 1
root = TreeNode(1)
root.right = TreeNode(2)
root.right.left = TreeNode(3)

sol = Solution()
result = sol.preorderTraversal(root)
print("Preorder traversal:", result)
